### ISSUE TITLE
HCAData

### ISSUE THREAD

**Issue by federicomarini** [Bioconductor/Contributions#937]
Update the following URL to point to the GitHub repository of
the package you wish to submit to _Bioconductor_

- Repository: https://github.com/federicomarini/HCAData

Confirm the following by editing each check box to '[x]'

- [x] I understand that by submitting my package to _Bioconductor_,
  the package source and all review commentary are visible to the
  general public.

- [x] I have read the _Bioconductor_ [Package Submission][2]
  instructions. My package is consistent with the _Bioconductor_
  [Package Guidelines][1].

- [x] I understand that a minimum requirement for package acceptance 
  is to pass R CMD check and R CMD BiocCheck with no ERROR or WARNINGS. 
  Passing these checks does not result in automatic acceptance. The 
  package will then undergo a formal review and recommendations for 
  acceptance regarding other Bioconductor standards will be addressed.

- [x] My package addresses statistical or bioinformatic issues related
  to the analysis and comprehension of high throughput genomic data.

- [x] I am committed to the long-term maintenance of my package. This
  includes monitoring the [support site][3] for issues that users may
  have, subscribing to the [bioc-devel][4] mailing list to stay aware
  of developments in the _Bioconductor_ community, responding promptly
  to requests for updates from the Core team in response to changes in
  _R_ or underlying software.

I am familiar with the essential aspects of _Bioconductor_ software
management, including:

- [x] The 'devel' branch for new packages and features.
- [x] The stable 'release' branch, made available every six
      months, for bug fixes.
- [x] _Bioconductor_ version control using [Git][5]
  (optionally [via GitHub][6]).

For help with submitting your package, please subscribe and post questions
to the [bioc-devel][4] mailing list.

[1]: https://bioconductor.org/developers/package-guidelines/
[2]: https://bioconductor.org/developers/package-submission/
[3]: https://support.bioconductor.org
[4]: https://stat.ethz.ch/mailman/listinfo/bioc-devel
[5]: http://bioconductor.org/developers/how-to/git/
[6]: http://bioconductor.org/developers/how-to/git/sync-existing-repositories/

**Comment by federicomarini** [comment]
As a note of interest, which I touched upon with Lori in a discussion on the bioc-community Slack channel:

I have seen that, for the recent hca application, Martin and Daniel are planning an HCABrowser package. Now, I don’t know the timeline for that, or what the full capabilities for this future package would be. In the meanwhile I think it is still worth having “as a standalone solution”, as I was also planning to have the datasets in this format as a playground to the `iSEE` package.

I'm not calling them out explicitly for the review, but I'm happy to hear from @mtmorgan and @dvantwisk

**Comment by federicomarini** [comment]
It worked @lshep, thank you! Still, I can't find out the reason of the warning message, which is namely `WARNING: directory 'HCAData/build' is empty` - since I am not using any such directory.

**Comment by lshep** [comment]
Not sure but I'd ignore it for now.  Let me know if you see any more abnormals

**Comment by federicomarini** [comment]
Ok, I'll let then the reviewers do their job 😉

**Comment by Liubuntu** [comment]
Hi @federicomarini , 

Thank you for your contribution to Bioconductor. This package is well-written
 and could be very useful. There are some minor issues from the technical review. 
Please seek to address most of them and comment back with any question or when 
ready for a 2nd look. Thanks!
 
## DESCRIPTION
  
- Please remove `lazyData: true`, This rarely proves to be a good
  thing. In our experience it only slows down the loading of packages
  with large data.

- Please remove the `Date:` field.

- `Description` field: Since the data returned by calling `HCAData()`
  is a `SingleCellExperiment` object, it would be helpful to add it
  into description field.

- For the same reason as above, you may want to `Depends:
  SingleCellExperiment` instead of just importing it, so that user
  could get automatic loading of that package and use functions
  directly without a manual `library(SingleCellExperiment)`. The
  package itself is light-weight and wouldn't make loading process
  heavy.

## R/

### HCAData.R

- Use `if(!is.null(dataset)) {...} else {` to avoid checking
  `is.null(dataset)` twice.
- SUGGESTED: This is more coding style thing but I just want to bring
  it into attention for a better tracking and maintenance of the code.
  Seems that you have been using `rdatapath` all through for rowData,
  colData and h5file. To reuse an R object, you may want to rename it
  as `datapath` to be more general, or use
  `rdatapath/cdatapath/h5datapath` separately.

## man/

### HCAData.md
- Please remove `library(SingleCellExperiment)` from example if you
  `Depends: SingleCellExperiment` in DESCRIPTION file.

## vignette

- Add a major section of `Introduction` (before `Installation`) to
  briefly introduce this package. You can use a relatively more
  detailed version of `Description` field inside `DESCRIPTION` file.
- Avoid a big chunk of code without enough explanation. In the section
  of `Processing a subset of the HCA bone marrow data`, as an example
  case study, please break it into smaller chunks with proper explanation
  for each processing step.
- The section`Saving computations` is not actually about "computation".
  You may want to rename it.
- some sentences are not smooth in reading. E.g., you may want to 
modify "instead storing the counts on disk..." as "instead, it stores ... and loads ...".

Cheers,
Qian

**Comment by federicomarini** [comment]
Hi Qian,

thank you a lot for the detailed review!

I'll work on the points you mentioned, and get back to you once all are addressed.

Federico

**Comment by federicomarini** [comment]
(out of curiosity: did you find out how the `HCAData/build` folder gets created and raises that warning in R CMD build?)

**Comment by federicomarini** [comment]
Qian, 

I am done with the first round. I addressed all points you mentioned, and I will go through them in detail in the following lines.

**DESCRIPTION**

> - Please remove lazyData: true, This rarely proves to be a good thing. In our experience it only slows down the loading of packages with large data.

I removed the line from the `DESCRIPTION` file

> Please remove the Date: field.

Same as above

> Description field: Since the data returned by calling HCAData() is a SingleCellExperiment object, it would be helpful to add it into description field.

Good point, I added that in the `Description`

> For the same reason as above, you may want to Depends: SingleCellExperiment instead of just importing it, so that user could get automatic loading of that package and use functions directly without a manual library(SingleCellExperiment). The package itself is light-weight and wouldn't make loading process heavy.

True, one will anyway have to play with these functions once the object is there, there is no point to explicitly have the user load it.

## R/

### HCAData.R

> Use if(!is.null(dataset)) {...} else { to avoid checking is.null(dataset) twice.

Nice catch. Edited accordingly

> SUGGESTED: This is more coding style thing but I just want to bring it into attention for a better tracking and maintenance of the code. Seems that you have been using rdatapath all through for rowData, colData and h5file. To reuse an R object, you may want to rename it as datapath to be more general, or use rdatapath/cdatapath/h5datapath separately.

Edited as in your suggestion.

## man/

### HCAData.md

> Please remove library(SingleCellExperiment) from example if you Depends: SingleCellExperiment in DESCRIPTION file.

I do now `Depend`, so this is gone. This is also gone in the unit tests.


## vignette

> Add a major section of Introduction (before Installation) to briefly introduce this package. You can use a relatively more detailed version of Description field inside DESCRIPTION file.

I added an intro section in the vignette, "Introducing the HCAData package"

> Avoid a big chunk of code without enough explanation. In the section of Processing a subset of the HCA bone marrow data, as an example case study, please break it into smaller chunks with proper explanation for each processing step.

I did break up the monster chunk in the vignette, and gave proper bits of explanation for each step.

> The section Saving computations is not actually about "computation". You may want to rename it.

Now it is called `## Saving the processed object`, more appropriately reflecting the content of the section.

> some sentences are not smooth in reading. E.g., you may want to modify "instead storing the counts on disk..." as "instead, it stores ... and loads ...".

Thank you for spotting this, sometimes my phrasing just gets overcomplicated. I edited that sentence, and also cleaned up some leftover commented lines throughout the vignette.

To sum up, thank you for the helpful suggestions. I am happy to proceed with the next round.
Side note: I see the build warning is still persisting, but could not track down why that is happening.

Cheers,
Federico

**Comment by Liubuntu** [comment]
Hi @federicomarini , 

Thank you for the prompt update! Now everything looks good for acceptance. For the warning, we'll just ignore it because it comes from `R CMD build`, there is nothing wrong on our end and nothing you could do to the package itself. We'll continue monitoring to see if it could pass someday later. 

Best,
Qian

**Comment by federicomarini** [comment]
Thank you for the good news!
Happy to contribute further to this project & community 👍

**Comment by mtmorgan** [comment]
The master branch of your GitHub repository has been added to Bioconductor's git repository.

To use the git.bioconductor.org repository, we need an 'ssh' key to associate with your github user name. If your GitHub account already has ssh public keys (https://github.com/federicomarini.keys is not empty), then no further steps are required. Otherwise, do the following:

1. [Add an SSH key to your github account][]
1. [Submit your SSH key to Bioconductor][]

See further instructions at

https://bioconductor.org/developers/how-to/git/

for working with this repository. See especially

https://bioconductor.org/developers/how-to/git/new-package-workflow/
https://bioconductor.org/developers/how-to/git/sync-existing-repositories/

to keep your GitHub and Bioconductor repositories in sync.

Your package will be included in the next nigthly 'devel' build (check-out from git at about 6 pm Eastern; build completion around 2pm Eastern the next day) at

https://bioconductor.org/checkResults/

(Builds sometimes fail, so ensure that the date stamps on the main landing page are consistent with the addition of your package). Once the package builds successfully, you package will be available for download in the 'Devel' version of Bioconductor using `BiocManager::install("HCAData")`. The package 'landing page' will be created at

https://bioconductor.org/packages/HCAData

If you have any questions, please contact the bioc-devel mailing list (https://stat.ethz.ch/mailman/listinfo/bioc-devel); this issue will not be monitored further.

[Add an SSH key to your github account]: https://help.github.com/articles/connecting-to-github-with-ssh/
[Submit your SSH key to Bioconductor]: https://git.bioconductor.org/BiocCredentials/
